src/core/lib/event_engine/workqueue.h

```diff
@@ -0,0 +1,238 @@
+// Copyright 2022 The gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#ifndef GRPC_CORE_LIB_EVENT_ENGINE_WORKQUEUE_H
+#define GRPC_CORE_LIB_EVENT_ENGINE_WORKQUEUE_H
+
+#include <grpc/support/port_platform.h>
+
+#include <stddef.h>
+
+#include <atomic>
+#include <deque>
+#include <utility>
+
+#include "absl/base/internal/spinlock.h"
+#include "absl/base/thread_annotations.h"
+#include "absl/cleanup/cleanup.h"
+#include "absl/synchronization/mutex.h"
+#include "absl/types/optional.h"
+
+#include "src/core/lib/gprpp/sync.h"
+#include "src/core/lib/gprpp/time.h"
+#include "src/core/lib/iomgr/exec_ctx.h"
+
+namespace grpc_event_engine {
+namespace experimental {
+
+// A fast work queue based lightly on an internal Google implementation.
+//
+// This uses atomics and lightweight spinlocks to access the most recent element
+// in the queue, making it fast for LIFO operations. Accessing the oldest (next)
+// element requires taking a mutex lock.
+template <typename T>
+class WorkQueue {
+ public:
+  static const grpc_core::Timestamp kInvalidTimestamp;
+
+  class Storage {
+   public:
+    Storage() = default;
+    Storage(T element, grpc_core::Timestamp enqueued)
+        : element_(element), enqueued_(enqueued) {}
+    ~Storage() = default;
+    // not copyable
+    Storage(const Storage&) = delete;
+    Storage& operator=(const Storage&) = delete;
+    // moveable
+    Storage(Storage&& other) noexcept
+        : element_(other.element_), enqueued_(other.enqueued_) {}
+    Storage& operator=(Storage&& other) noexcept {
+      std::swap(element_, other.element_);
+      std::swap(enqueued_, other.enqueued_);
+      return *this;
+    }
+    grpc_core::Timestamp enqueued() const { return enqueued_; }
+    T&& TakeElement() { return std::move(element_); }
+
+   private:
+    T element_;
+    grpc_core::Timestamp enqueued_ = kInvalidTimestamp;
+  };
+
+  WorkQueue() = default;
+
+  // Returns whether the queue is empty
+  bool Empty() const {
+    return (most_recent_element_enqueue_timestamp_.load(
+                std::memory_order_relaxed) == kInvalidTimestamp &&
+            oldest_enqueued_timestamp_.load(std::memory_order_relaxed) ==
+                kInvalidTimestamp);
+  }
+
+  // Returns the number of elements in the queue
+  // TODO(hork): this is an expensive method. consider caching the size in an
```

Comment:
Probably not worth caching: we shouldn't be calling much outside of tests or some very infrequent stats collection.

Reply:
ack, I replaced it with a comment about the behavior.